From 1f7972c8893ff190ea5786b35e68e8a0ec1fef55 Mon Sep 17 00:00:00 2001
From: Eric Gillespie <epg@pretzelnet.org>
Date: Thu, 11 Dec 2025 11:58:02 -0600
Subject: [PATCH 36/74] Report failure of 'create --command' command

When we can't even run a child process at all (system(3) returns -1),
throw runtime_error_with_errno without taking the post snapshot.

When the child is killed or exits with non-zero status, report that to
stderr and exit(EXIT_FAILURE) .
---
 client/snapper/cmd-create.cc | 16 +++++++++++++++-
 doc/snapper.xml.in           |  5 ++++-
 2 files changed, 19 insertions(+), 2 deletions(-)

diff --git a/client/snapper/cmd-create.cc b/client/snapper/cmd-create.cc
index 75f67436..65cbaeec 100644
--- a/client/snapper/cmd-create.cc
+++ b/client/snapper/cmd-create.cc
@@ -21,6 +21,8 @@
  */
 
 
+#include <sys/wait.h>
+
 #include <iostream>
 
 #include <snapper/AppUtil.h>
@@ -193,10 +195,22 @@ namespace snapper
 
 	    case CreateType::PRE_POST: {
 		snapshot1 = snapper->createPreSnapshot(scd, report);
-		system(command.c_str());
+		int const status = system(command.c_str());
+		if (status == -1) {
+		    throw runtime_error_with_errno("open failed", errno);
+		}
 		snapshot2 = snapper->createPostSnapshot(snapshot1, scd, report);
 		if (print_number)
 		    cout << snapshot1->getNum() << ".." << snapshot2->getNum() << endl;
+		if (WIFEXITED(status)) {
+		    int const exit_status = WEXITSTATUS(status);
+		    if (exit_status != 0) {
+			// TODO(epg): Can we have snapper itself exit with this status?
+			SN_THROW(Exception(sformat("%s exited %d", command.c_str(), exit_status)));
+		    }
+		} else {
+		    SN_THROW(Exception(sformat("%s killed with %d", command.c_str(), WTERMSIG(status))));
+		}
 	    } break;
 	}
     }
diff --git a/doc/snapper.xml.in b/doc/snapper.xml.in
index 613e07d0..2b4a793c 100644
--- a/doc/snapper.xml.in
+++ b/doc/snapper.xml.in
@@ -535,7 +535,10 @@
 	    <varlistentry>
 	      <term><option>--command</option> <replaceable>command</replaceable></term>
 	      <listitem>
-		<para>Create a pre and post snapshot and run command in between.</para>
+		<para>Create a pre and post snapshot and run command in between.  If snapper is
+		unable to execute the command at all, it does not take the post snapshot.  If the
+		command exits with a non-zero exit code, snapper takes the post snapshot and exits 1
+		to indicate the failure.</para>
 	      </listitem>
 	    </varlistentry>
 	    <varlistentry>
-- 
2.43.0

